filters.go

```diff
@@ -49,6 +77,40 @@ func (p *Pipe) Concat() *Pipe {
 	return p.WithReader(io.MultiReader(readers...))
 }
 
+// Dirname reads a list of filenames from the pipe, one per line, and returns
+// a pipe which contains only the directory names of each filename.
+//
+// If a line is empty, Dirname will return a '.' for that line
+// Trailing slashes are removed, unless Dirname returns the root folder.
+func (p *Pipe) Dirname() *Pipe {
+	// do we have anything to do?
+	if p == nil || p.Error() != nil {
+		return p
+	}
+
+	return p.EachLine(func(line string, out *strings.Builder) {
+		// special case:
+		//
+		// filepath.Dir() does not handle trailing slashes correctly
+		// we have to strip the trailing slash ourselves
+		if len(line) > 1 && line[len(line)-1:] == "/" {
+			line = line[0 : len(line)-1]
+		}
+		dirname := filepath.Dir(line)
+
+		// special case:
+		//
+		// filepath.Dir() does not preserve any './' at the front of
+		// a path
+		if len(dirname) > 1 && len(line) > 1 && line[:2] == "./" {
+			dirname = "./" + dirname
```

Comment:
To save a string copy, maybe `out.WriteString("./")` is better here?

Reply:
Keeping the `out.WriteXXX()` calls co-located in a single block normally avoids maintenance issues over the lifetime of a piece of code. I'd stick with the string copy.

Reply:
That's fair enough.